### ISSUE TITLE
Endpoint to list nullifiers and basic E2E infrastructure

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-node#10]
We should implement components for handling RPC requests end-to-end. This includes:

- RPC component.
- Store component.
- Database.

The idea is to have a single endpoint on the RPC side. When we make a request to this endpoint, RPC forwards the request to the store, and store retrieves the data from the database/internal memory and returns this.

For the PoC endpoint we could pick `CheckNullifiers` endpoint which would accept a list of nullifiers and return a list of consumed nullifiers together with their authentication paths against the current nullifier tree root.

On the store side, we'd need to have the following:
- A database with a single `nullifier` table. The table would have just 2 columns:
  - `nullifier` (blob, unique primary key).
  - `block_num` (u32) - block number at which the nullifier was created.
- An in-memory Sparse Merkle tree (our TSMT) for all nullifiers. This tree would be used to retrieve Merkle paths for the returned nullifiers.
  - The store would need to build this tree at start-up from the data in the `nullifier` table.

**Comment by bobbinth** [comment]
Actually, I think the response should include all nullifiers (together with authentication paths). This will be needed to prove which nullifiers have been consumed.

**Comment by hackaugusto** [comment]
I wondered if the available libraries would have a big impact on performance, when looking over the available ones, specially since tokio_rusqlite uses a dedicated thread for interacting with the DB while deadpool uses tokio's threadpool, so I wrote a few simple benchmarks:

<details><summary>Single threaded</summary>

```rust
use rusqlite::Connection;
use std::time::Instant;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // use std::fs;
    // let _ = fs::remove_file("test.sqlite3");

    let start = Instant::now();
    let conn = Connection::open("test.sqlite3")?;

    conn.execute(
        "CREATE TABLE IF NOT EXISTS nullifier (
            id INTEGER PRIMARY KEY,
            nullifier BLOB,
            path BLOB
        );",
        [],
    )?;

    let nullifier: [u8; 256] = [0x01; 256];
    let path: [u8; 2048] = [0x02; 2048];

    for _ in 0..1024 {
        conn.execute(
            "INSERT INTO nullifier (nullifier, path) VALUES (?1, ?2)",
            (&nullifier, &path),
        )?;
    }

    println!("Elapsed: {:.2?}", start.elapsed());
    Ok(())
}
```

</details>

<details><summary>Multi threaded</summary>

```rust
use rusqlite::Connection;
use std::thread;
use std::time::Instant;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // use std::fs;
    // let _ = fs::remove_file("test.sqlite3");

    let start = Instant::now();
    let conn = Connection::open("test.sqlite3")?;

    conn.execute(
        "CREATE TABLE IF NOT EXISTS nullifier (
            id INTEGER PRIMARY KEY,
            nullifier BLOB,
            path BLOB
        );",
        [],
    )?;

    let nullifier: [u8; 256] = [0x01; 256];
    let path: [u8; 2048] = [0x02; 2048];

    let r1 = thread::spawn(move || {
        let conn = Connection::open("test.sqlite3").unwrap();
        for _ in 0..512 {
            conn.execute(
                "INSERT INTO nullifier (nullifier, path) VALUES (?1, ?2)",
                (&nullifier, &path),
            )
            .unwrap();
        }
    });
    let r2 = thread::spawn(move || {
        let conn = Connection::open("test.sqlite3").unwrap();
        for _ in 0..512 {
            conn.execute(
                "INSERT INTO nullifier (nullifier, path) VALUES (?1, ?2)",
                (&nullifier, &path),
            )
            .unwrap();
        }
    });

    r1.join().unwrap();
    r2.join().unwrap();

    println!("Elapsed: {:.2?}", start.elapsed());
    Ok(())
}
```
</details>

<details><summary>Async w/ deadpool</summary>

```rust
use deadpool_sqlite::{Config, Runtime};
use std::time::Instant;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // use std::fs;
    // let _ = fs::remove_file("test.sqlite3");

    let start = Instant::now();
    let cfg = Config::new("test.sqlite3");
    let pool = cfg.create_pool(Runtime::Tokio1).unwrap();
    let conn = pool.get().await.unwrap();

    conn
        .interact(|conn| {
            conn.execute(
                "CREATE TABLE IF NOT EXISTS nullifier (
                    id INTEGER PRIMARY KEY,
                    nullifier BLOB,
                    path BLOB
                );",
                [],
            )
        })
        .await??;

    let nullifier: [u8; 256] = [0x01; 256];
    let path: [u8; 2048] = [0x02; 2048];

    for _ in 0..1024 {
        conn
            .interact(move |conn| {
                conn.execute(
                    "INSERT INTO nullifier (nullifier, path) VALUES (?1, ?2)",
                    (&nullifier, &path),
                )
            })
            .await??;
    }

    println!("Elapsed: {:.2?}", start.elapsed());
    Ok(())
}
```
</details>

<details><summary>Async w/ tokio_rusqlite</summary>

```rust
use rusqlite::Result;
use std::time::Instant;
use tokio_rusqlite::Connection;

#[tokio::main]
async fn main() -> Result<()> {
    // use std::fs;
    // let _ = fs::remove_file("test.sqlite3");

    let start = Instant::now();
    let conn = Connection::open("test.sqlite3").await?;
    // let conn = Connection::open_in_memory().await?;

    conn
        .call(|conn| {
            conn.execute(
                "CREATE TABLE IF NOT EXISTS nullifier (
                    id INTEGER PRIMARY KEY,
                    nullifier BLOB,
                    path BLOB
                );",
                [],
            )
        })
        .await?;

    let nullifier: [u8; 256] = [0x01; 256];
    let path: [u8; 2048] = [0x02; 2048];

    for _ in 0..1024 {
        conn
            .call(move |conn| {
                conn.execute(
                    "INSERT INTO nullifier (nullifier, path) VALUES (?1, ?2)",
                    (&nullifier, &path),
                )
            })
            .await?;
    }

    println!("Elapsed: {:.2?}", start.elapsed());
    Ok(())
}
```
</details>

The performance is round about the same and differences can be assigned to noise. So I think using deadpool which has the most users and its co-mainted by tokio is the best option (even though the code is a bit more complex and there is no dedicated background thread for it)

**Comment by bobbinth** [comment]
I'm actually fine with either `deadpool_sqlite` or `tokio_rusqlite`. Database access logic will probably be fairly well encapsulated (I'm imagining something like a `Db` struct which exposes specialized `async` methods) and switching from one to another shouldn't be too difficult (and we can even implement our own connection pool in the future, if needed).

Unrelated comment: I noticed that nullifier table is defined as `(nullifier, path)` - is this for benchmarking only? Basically, in the real implementation, we shouldn't store the path in the table.

**Comment by hackaugusto** [comment]
> is this for benchmarking only? 

yes

**Comment by bobbinth** [comment]
Closed by #14 and related PRs.
